pkg/apis/batch/validation/validation.go

```diff
@@ -518,6 +518,16 @@ func validateJobStatus(job *batch.Job, fldPath *field.Path, opts JobStatusValida
 			allErrs = append(allErrs, field.Invalid(fldPath.Child("completionTime"), status.CompletionTime, "completionTime cannot be set before startTime"))
 		}
 	}
+	if opts.RejectFailedJobWithoutFailureTarget {
```

Comment:
I'm not sure we should be validating this (same for SuccessCriteriaMet).

You could imagine a job controller that doesn't have these conditions but still waits for everything to finish before setting the terminal condition.

Reply:
Yeah, I'm not fully sure either, but this was the suggestion: https://github.com/kubernetes/enhancements/pull/4657#discussion_r1620853394. 

I think it is reasonable to assume that any Job controller implementation should communicate the fate of the Job as soon as possible, especially since now setting the terminal condition can be delayed. So, adding the validation we make sure the developers are aware of the conditions and would not introduce custom conditions for this purpose.

Reply:
I would have preferred to save on complexity of validation code, but since it's already approved in the KEP, I'm ok with it.

Reply:
It's also important for consumers of the Job API, that all the job controller implementations communicate status in the same way with the same meanings.  Accepting some complexity in validation to ensure that API consumers can be assured of congruent semantics between implementations is net win for the community.